ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HeaderValidation.hs

```diff
@@ -499,6 +510,75 @@ deriving instance StandardHash blk => Eq   (TipInfoIsEBB blk)
 deriving instance StandardHash blk => Show (TipInfoIsEBB blk)
 deriving instance StandardHash blk => NoThunks (TipInfoIsEBB blk)
 
+{-------------------------------------------------------------------------------
+  Header with time
+-------------------------------------------------------------------------------}
+
+-- | A header paired with the time of the slot that it inhabits.
+--
+-- Note that the header's slot was translated to this time (in the ChainSync
+-- client) according to the header's chain. This clarification may be helpful,
+-- since it's possible that some other chain would translate that same slot to
+-- a different time.
+data HeaderWithTime blk = HeaderWithTime {
+    hwtHeader           :: !(Header blk)
+  , hwtSlotRelativeTime :: !RelativeTime
+  }
+  deriving (Generic)
+
+deriving stock instance (Eq (Header blk))
+                      => Eq (HeaderWithTime blk)
+deriving stock instance (Show (Header blk))
+                      => Show (HeaderWithTime blk)
+deriving anyclass instance (NoThunks (Header blk))
+                         => NoThunks (HeaderWithTime blk)
+
+type instance HeaderHash (HeaderWithTime blk) = HeaderHash (Header blk)
+
+instance ( Show (HeaderHash blk)
+         , Eq (HeaderHash blk)
+         , Ord (HeaderHash blk)
+         , Typeable (HeaderHash blk)
+         , NoThunks (HeaderHash blk)
+         ) => StandardHash (HeaderWithTime blk)
+
+instance ( HasHeader (Header blk)
+         , StandardHash (HeaderWithTime blk)
+         , Typeable blk
```

Comment:
This is the root cause of all the additional `Typeable blk` contexts in this commit. It's required to satisfy the superclass on the (upstream) `HasHeader`. I don't see how to avoid it, sadly.